File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = -10**18
        def solve(node):
            nonlocal ans
            if node is None:
                return 0
            
            l = solve(node.left)
            r = solve(node.right)

            ans1 = l + r + node.val
            ans2 = max(l,r) + node.val
            ans3 = node.val

            ans = max([ans,ans1,ans2,ans3])
            return max(ans2,ans3)
        
        solve(root)
        return ans 
        

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, int &ans){
        if (node == NULL) return 0;
        int leftsum = max(0, dfs(node->left,ans)); //max (0,leftsum) to ignore negatives
        int rightsum = max(0, dfs(node->right, ans)); //max (0,rightsum) to ignore negatives
        ans = max(ans, node->val+leftsum+rightsum);
        return node->val + max(leftsum,rightsum);
    }
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        dfs(root, ans);
        return ans;
    }
};